src/coord/src/catalog.rs

```diff
@@ -1250,6 +1251,65 @@ impl Catalog {
         Ok(temporary_ids)
     }
 
+    /// Insert timestamp bindings into SQLite, and ignores duplicate timestamp bindings.
+    /// TODO: we intentionally ignore duplicates because BYO sources can send multiple
+    /// copies of the same timestamp.
+    pub fn insert_timestamps(
+        &mut self,
+        timestamps: &[(GlobalId, String, Timestamp, i64)],
+    ) -> Result<(), Error> {
+        let mut storage = self.storage();
+        let tx = storage.transaction()?;
+
+        for (sid, pid, ts, offset) in timestamps.iter() {
+            tx.insert_timestamp(sid, pid, *ts, *offset)?;
+        }
+        tx.commit()?;
+
+        Ok(())
+    }
+
+    /// Read all available timestamp bindings for a source
+    ///
+    /// Returns its output sorted by (partition, timestamp)
+    pub fn load_timestamps(
+        &mut self,
+        source_id: GlobalId,
+    ) -> Result<Vec<(PartitionId, Timestamp, MzOffset)>, Error> {
+        let mut storage = self.storage();
+        let tx = storage.transaction()?;
+
+        let ret = tx.load_timestamps(source_id)?;
+        tx.commit()?;
+
+        Ok(ret)
+    }
+
+    /// Delete all timestamp bindings for a source
+    pub fn delete_timestamps(&mut self, source_id: GlobalId) -> Result<(), Error> {
+        let mut storage = self.storage();
+        let tx = storage.transaction()?;
+
+        tx.delete_timestamps(source_id)?;
+        tx.commit()?;
+
+        Ok(())
+    }
+
+    pub fn compact_timestamps(
```

Comment:
There's a lot of no docs in this file. This happens to be the first one, and seems like a likely hard to understand method.

Reply:
Sorry, github ui issue. This method has no docs, and there are a lot of no docs in `catalog/storage.rs`.